### PR TITLE
Update theme design after layout merge called get theme

### LINKED ISSUE #30656: \Magento\Framework\View\Model\Layout\Merge stores the theme at construct time
### Preconditions (*)

Magento 2.3.5

### Steps to reproduce (*)

1. Set a specific theme for a Product
2. Add a widget expected to be displayed on all the product pages using the specific theme above
3. Load the page of the Product that you set a specific theme at step 1

### Expected result (*)

The specific theme is used _and_ the widget is displayed.

### Actual result (*)

The specific theme is used but the widget is _not_ displayed.

### Explanation

`\Magento\Framework\View\Model\Layout\Merge` stores the theme at construct time.

This creates a major issue with the rendering process, when one of the actor wants to change the theme - like for example the Magento_Catalog module. Since the theme is stored at construct time, would it be changed later, the `Merge` instance would always use the former theme stored at construct time to fetch layout updates.

The most spectacuar issues it creates is when trying to assign widgets for catalog layout handles (like catalog_product_view). Since the `Merge` instance always uses the default theme to fetch the layout updates, it never gets the chance to fetch the ones that are assigned to the widget if the theme of the product is _not_ set to the default theme of the website. The only way for the widget to get grabbed is for it to be assigned to the default theme of the website, whatever the custom theme applied to the products.

Fortunately it should be easy to write a plugin that intercepts `getTheme` from the `Merge` instance to return the current theme instead of the one set at construct time.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

Plugin aroundGetDbUpdateString
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Widget/Model/ResourceModel/Layout/Plugin.php#L41

Widget update Layout called from plugin above
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Widget/Model/ResourceModel/Layout/Update.php#L63

plugin module widget `aroundGetDbUpdateString` now retrieve correct theme instance instead get theme instance from construction(when layout merge init theme) 
https://github.com/magento/magento2/blob/5c31d9aead63106ff72a22b50458ef2112ffdff6/lib/internal/Magento/Framework/View/Model/Layout/Merge.php#L235

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30656

### Manual testing scenarios (*)

Pre-config
Setup theme config default for store => luma theme
Scenario1
Product named "Joust Duffle Bag"
1. Go to admin -> Change Schedule Design Update : New theme: Choose blank theme
2. Create widget show products list type for blank theme and display only in product detail for "Joust Duffle Bag"
2.1 Type: Catalog Products List
2.2 Design Package/Theme: Blank theme
2.3 Assign to Store Views: All Store Views
2.4 Layout Updates: All product types
Specific Products: choose "Joust Duffle Bag"
Contianer: Main Container Bottom
Widget options: Choose default to make sure all critieria meet and valid widget to show
3. Save widget
4. Clear cache
5. Go to homepage first (Verify theme blank applied) OK it's work fine
6. Go to frontend visit page product "Joust Duffle Bag". Theme blank applied ok and expect widget created previously should visible
7.After page loaded see results. Theme blank show but widget not visible

### Expected results
page details of `Joust Duffle Bag` should applied blank theme (theme blank) not luma. Set xdebug breakpoint may help for easier verify

Widget should show and
Theme blank applied (Correct display)

Widget Catalog Product List should show in area `Main Content Top` of page details `Joust Duffle Bag` only

### Actually results (before fixes)
Widget Catalog Product List not visible on details page of `Joust Duffle Bag` (luma theme instance still used)

Widget not show (because magento still get theme luma default from initialized, Actually in the update layout theme applied use should be blank)
Theme blank applied (Correct display)

### Questions or comments

CC: @ericmorand @ihor-sviziev @kandy @orlangur

Test
@engcom-Alfa @engcom-Charlie 
Need coverage with MFTF will be very good
Add test for plugin widget non-coverage by any tests
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Widget/Model/ResourceModel/Layout/Plugin.php

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Integration Tests, Unit Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Seems like this fix should be covered with MFTF tests, unit test isn't enough here

**Comment by mrtuvn** [comment]
agreed. Current i not sure how to start with tests for such case

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento give me test instance

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

File: app/code/Magento/Widget/Model/ResourceModel/Layout/Plugin.php [hunk 1]
```diff
@@ -38,6 +51,8 @@ public function aroundGetDbUpdateString(
         \Closure $proceed,
         $handle
     ) {
-        return $this->update->fetchUpdatesByHandle($handle, $subject->getTheme(), $subject->getScope());
+        /** @var ThemeInterface $theme */
+        $theme = $this->themeResolver->get();
```
Comment (mrtuvn) [inline app/code/Magento/Widget/Model/ResourceModel/Layout/Plugin.php]: Retrieve instance of a theme currently used in an area

**Comment by ihor-sviziev** [comment]
@mrtuvn pls let me know when we can review this PR

**Comment by mrtuvn** [comment]
yes i have updated last change

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Framework/View/Model/Layout/MergeTest.php [hunk 2]
```diff
@@ -77,10 +130,24 @@ protected function setUp(): void
...
                 'theme' => $theme,
+                'cacheSuffix' => '',
                 'layoutCacheKey' => $this->layoutCacheKeyMock,
+                'serializer' => $this->serializer,
+                'cacheLifetime' => self::STUB_DEFAULT_CACHE_LIFETIME
             ]
         );
+
+
     }
```
Comment (mrtuvn) [inline dev/tests/integration/testsuite/Magento/Framework/View/Model/Layout/MergeTest.php]: Do we need pass all arguments from constructor in test ? Found some other place only have fewer arguments
Example 
dev/tests/integration/testsuite/Magento/PageCache/Model/Layout/MergeTest.php
https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/PageCache/Model/Layout/MergeTest.php
CC: @ihor-sviziev @paliarush

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests CE

**Comment by mrtuvn** [comment]
@magento run Integration Tests

File: dev/tests/integration/testsuite/Magento/Framework/View/Model/Layout/MergeTest.php (hunk 2 above)
Comment (ihor-sviziev) [reply dev/tests/integration/testsuite/Magento/Framework/View/Model/Layout/MergeTest.php]: Basically, you don't need to pass all the arguments in the integration test. You need to pass only mocks if you have any. Suppose the original class is OK for you - no need to pass it to the constructor.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Integration tests are failing, please check them

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
closing this PR seem it's introduced more regression issues

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Integration Tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests CE

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Unit Tests, Static Tests, Integration Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Unit Tests, Static Tests, Integration Tests

**Comment by mrtuvn** [comment]
May will close this one since core team already fixed this issue
https://github.com/magento/magento2/commit/76ff97efcc6352019f3775e2e7f62260fb8b0679#diff-d036d544db88cf91738bb3b56dbe89e156d4043e4cb7062af1eda9d20911d672R196

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Reopen this one since problem may able reproduce in difference case

**Comment by mrtuvn** [comment]
@magento run Unit Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Unit Tests, Static Tests, Integration Tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Hi @mrtuvn 

Thank you for your contribution.

I will work on MFTF test for this issue.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
I have created and pushed new MFTF test case for the mentioned scenario in ticket description.

Before proceeding further on failed test cases in build I need some information. 

I have tried to reproduce this issue on magento 2.4-develop and its not reproducible. 

@mrtuvn , as per your [comment](https://github.com/magento/magento2/pull/31929#issuecomment-883558184) can you please tell us, what are the other reproducible scenarios?

If those scenarios are still exists then will go with checking the build failure test cases. If they are not reproducible anymore then we can proceed with either one of the way as below:

1. We will remove the fix from this issue and only go with the MFTF test case related commit in this ticket(as its not covered yet). We also have to edit the description of ticket accordingly in this case.

2. Or we can raise new PR with MFTF test case commit and close this issue.

@mrtuvn please let us know so that we can proceed further on this issue.

Thank you in advance!

**Comment by mrtuvn** [comment]
Can you verify this case ?

Scenario1
Product named "Joust Duffle Bag"

Go to admin -> Change Schedule Design Update : New theme: Choose blank theme
Create widget show products list type for blank theme and display only in product detail for "Joust Duffle Bag"
2.1 Type: Catalog Products List
2.2 Design Package/Theme: Blank theme
2.3 Assign to Store Views: All Store Views
2.4 Layout Updates: All product types
Specific Products: choose "Joust Duffle Bag"
Contianer: Main Container Bottom
Widget options: Choose default to make sure all critieria meet and valid widget to show
Save widget
Clear cache
Go to homepage first (Verify theme luma applied) OK it's work fine
Go to frontend visit page product "Joust Duffle Bag". Theme blank applied ok and expected widget created previously should visible (show)
7.Go to backend  > Edit product Joust Duffle Bag again
8. Now we change schedule date udpate to future
9. Save product and clear cache
10. Back to product Joust Duffle Bag details and visual content

Expected Luma theme

Actually Blank theme and widget still show

**Comment by mrtuvn** [comment]
Thanks for your work. Let's wait owner re-verify problem at linked issue. If we can't reproduce i believe we dont have to any additional changes

**Comment by engcom-Charlie** [comment]
Hi @mrtuvn,

We are waiting for the updates which you have mentioned in [comment](https://github.com/magento/magento2/pull/31929#issuecomment-923994782). 

Thank you!

**Comment by mrtuvn** [comment]
Closed this since i can't reproduce this case in latest code

**Pull request closed without merging**